### PR TITLE
azure: add worker plugin for spot instances

### LINKED ISSUE #239: Add plugin for preemptible instances.
Hi !

Following some discussions in dask/distributed#4421. I've implemented a worker plugin that calls azure metadata service to check for node eviction in AKS and try to shut down workers as cleanly as possible. @jacobtomlinson suggested I contribute it to this repo in https://github.com/dask/distributed/issues/4421#issuecomment-760129128 somewhere in `dask_cloudprovider.azure.utils`.

Before opening an MR I would like some pointers on where to put the actual code and the associated documentation.
Thank you.

### PR DESCRIPTION
**Pull request by Timost**
Azure spot instances may be preempted anytime which might lead to data
losses on the dask cluster. This commit adds a worker plugin that tries
to close worker gracefully before the VM is preempted. To do so, it
polls azure metadata service to listen to premption events
notifications.

fixes #239

### CODE REVIEW

**Comment by Timost** [comment]
I couldn't build the docs on my computer (import errors due to the submodule approach I think, so I'm not sure how it renders.

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,5 @@
+aiohttp>=3.7.3
 dask>=2021.01.1
 distributed>=2021.01.1
 jinja2
+tornado >= 5
```
Comment (jacobtomlinson) [inline requirements.txt:5]:
```suggestion
tornado>=5
```

File: dask_cloudprovider/azure/utils.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        termination_events=None,
+        termination_offset_minutes=0,
+    ):
+        self.callback = None
+        self.loop = None
+        self.worker = None
+        self.poll_interval_s = poll_interval_s
+        self.metadata_url = (
+            metadata_url
+            or "http://169.254.169.254/metadata/scheduledevents?api-version=2019-08-01"
```
Comment (jacobtomlinson) [inline dask_cloudprovider/azure/utils.py:75]: This should probably be a constant defined at the top of the file.

File: dask_cloudprovider/azure/utils.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        Extra offset to apply to the premption date. This may be negative, to start
+        the gracefull shutdown before the ``NotBefore`` date. It can also be positive, to
+        start the shutdown after the ``NotBefore`` date, but this is at your own risk.
+
+        Defaults to ``0``
+
+    Examples
+    --------
+
+    Let's say you have cluster and a client instance. For example using ``KubeCluster``
```
Comment (jacobtomlinson) [inline dask_cloudprovider/azure/utils.py:50]: Let's make this a full reference to span the two sets of docs. Will need to add dask-kubernetes to the intersphinx mapping.

https://github.com/dask/dask-cloudprovider/blob/dffd81d3e4ab6feb5378ffa0d9952f3c7da207f7/doc/source/conf.py#L194-L198

```suggestion
    Let's say you have cluster and a client instance. For example using :class:`dask_kubernetes.KubeCluster`
```

File: dask_cloudprovider/azure/utils.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+logger = logging.getLogger(__name__)
+
+
+class AzurePreemptibleWorkerPlugin(WorkerPlugin):
+    """A worker plugin for azure spot instances
+
+    This worker plugin will poll azure's metadata service for preemption notifications.
+    When a node is preempted, the plugin will attempt to shutdown gracefully all workers
+    on the node.
+
```
Comment (jacobtomlinson) [inline dask_cloudprovider/azure/utils.py:19]: It would be good to note here that this plugin can be used on any worker running on Azure Preemtible VMs, not just ones created by `dask-cloudprovider`.

**Review by jacobtomlinson — COMMENTED** [review]
Thanks for this! A few comments.

Your docs probably wont build locally if you don't have all the optional dependencies installed. You can install everything with `pip install -e .[all]`. But we also have the Read The Docs PR preview builds turned on, so if you look at the CI checks you will find a link to a preview of the docs.

**Comment by Timost** [comment]
Thank you for the feedback. I installed everything and now the `make html` works ( with a lot of warnings though). :heart: for the docs in the CI.
I'll make the changes and update the PR !

**Comment by TomAugspurger** [comment]
This is neat to see. If we're able to, it'd be great to determine what parts are specific to Azure VMs, and what applies to other Azure services like AKS. If this works with AKS as well, then I think it'd be appropriate to include the plugin in `distributed` instead of `dask-cloudprovider`.

**Comment by Timost** [comment]
I'm actually using this with AKS (with vmss backed spot node pools) not Azure VMs, in fact I haven't tested it with Azure VMs.

If the consensus is to put this in `distributed` I don't mind moving the code over there. I'm don't have a strong opinion on this, except that `dask-cloudprovider` is quite new and every `dask` user might not be aware it exists.

**Comment by jacobtomlinson** [comment]
> If this works with AKS as well, then I think it'd be appropriate to include the plugin in distributed instead of dask-cloudprovider so that dask-gateway and dask-kubernetes can benefit from it as well.

It should work on everything preemptible in Azure.

We did discuss this over in distributed, but felt that cloudprovider is the appropriate place for cloud specific things. There is nothing stopping you from making use of this with other libraries. 

Perhaps we should consider adding references to it in other docs.

**Comment by TomAugspurger** [comment]
Mmm, was that discussed in https://github.com/dask/distributed/issues/3300?

I'd find it a bit strange to install dask-cloudprovider if you're *only* using it for this plugin (if you're using dask-kubernetes, say).

While it *is* cloud-specific, I think that since multiple sub-projects would get value out of it, `distributed` would make a good home, since each of the sub-projects depends on distributed.

**Comment by Timost** [comment]
It was discussed in https://github.com/dask/distributed/issues/4421. Sorry it's kind of hard to follow. The original issue ended up raising several other issues on different projects.

I understand both points of view, I think `dask-cloudprovider` has several submodules that can be installed separately. So, another solution could be putting the plugin in a separate module that can be installed independently of the rest of `dask-cloudprovider`. What do you think ? Maybe that's too much modules...

**Comment by jacobtomlinson** [comment]
@TomAugspurger I see where you are coming from.

My thinking here was that these plugins may have cloud dependencies (although I don't think this one does).

So installing `dask-cloudprovider` (or specifically `pip install dask-cloudprovier[azure]`) to get cloud specific libraries and dependencies makes sense.

The cloud vendor dependencies can be very heavy and we install them here only as extras. I would like to avoid that code going into distributed.

**Comment by TomAugspurger** [comment]
Yeah, the dependency point is strong. I agree that when there are cloud-specific dependencies, it's best to have them in a package that depends on them rather than relying on runtime checks. Since this plugin doesn't do anything with azure.*, my thought is it would be better suited for distributed (aiohttp would need to be replaced with tornado.httpclient).

But I don't know what other cloud-provider specific plugins are available in dask-cloudprovider, and it would be a bit strange to have some in distributed and some in dask-cloudprovider, depending on their implementation details.

**Comment by Timost** [comment]
> (aiohttp would need to be replaced with tornado.httpclient).

Thank you for pointing this out, I'll keep it in mind.

> But I don't know what other cloud-provider specific plugins are available in dask-cloudprovider, and it would be a bit strange to have some in distributed and some in dask-cloudprovider, depending on their implementation details.

AFAIK there are no other cloud-provider specific plugins in `dask-cloudprovider` for now.

I'll soon get the opportunity to run the same pipeline on google's GKE, so I thought I might mention it here as it would have triggered writing an equivalent plugin for gcp. But it seems google's preemption notification is done entirely differently (https://cloud.google.com/compute/docs/instances/preemptible#preemption-process).

**Comment by TomAugspurger** [comment]
Don't let me comments about where this should live derail the implementation here. If we decide that distributed is a better home in the future then we can make the migration then.

**Comment by Timost** [comment]
PTAL

**Comment by jacobtomlinson** [comment]
Ok this looks great.

I'm going to get this merged in here now and also open an issue on Distributed to discuss upstreaming it. I'll raise it at the Dask maintainers meeting too.

**Comment by Timost** [comment]
:+1: Thank you !
